### ISSUE TITLE
Some places appear to be considered a wall in the navmesh when they shouldn't be

### ISSUE THREAD

**Issue by Hozu** [DarkstarProject/darkstar#2837]
Client Version (type /ver in game) :
30160203_0

Server Version (type @revision in game) :
14a1e3d

Source Branch (master/stable) :
master

Additional Information (Steps to reproduce/Expected behavior) :
Stairs in various zones (Dynamis - Bastok (AH), Dynamis - Xarcabard (the structures the NMs are on), and The Eldieme Necropolis (the areas next to the plates that toggle the doors). Before line of sight was implemented, mobs had an extremely difficult, if not impossible time pathing up/down the stairs, but with line of sight it's impossible to pull the mobs without going up to them.

Another issue is the bushes that players can traverse through in Caedarva Mire by where Khimaira spawns (and presumably the similar ones in Wajaom Woodlands in the southern part of the main map) - there was an imp that someone managed to get pulled into there and it could not be attacked, nor would it aggro even when I was inside the mob's model.

**Comment by Fiocitrine** [comment]
Also, Castle O where the okote NM spawns. If you aggro a mob and it goes under the bridge it'll just move back and forth underneath the bridge. Doing this, it will not be able to attack you but still be in an aggro state.

**Comment by aizenmyou** [comment]
I also experienced similar on the same server Hozu is on. In Ro'Maeve if you stand on the edge of the fountain (Not the part up the stairs but the actual light gray edge as it meets the dark gray of the normal floor) I could not use ranged attack as RNG (Got yellow system message)

I've also see this with the small rocks you can walk over in the same zone when being behind them.

**Comment by gedads** [comment]
"but with line of sight it's impossible to pull the mobs without going up to them."
Maybe removing the stairs a add a plane surface instead could help, stairs often have some areas left aside in recast

**Comment by Hozu** [comment]
What do stairs look like now? Small walls for each riser? If so, what would be the implications of ignoring really short walls?

**Comment by gedads** [comment]
for caedarva mire, the imps went through the bush on my server in the path to khimaira but it can be stuck i think .they 're easy to remove if needed. (http://imgur.com/3aaSxKe).
actually the stairs are like that: http://imgur.com/NI3p6Vu

**Comment by Hozu** [comment]
Yeah I got the imps to go through the bush, but when they were in it after deaggroing, they wouldn't aggro to anyone and they couldn't be acted on (other than by melee).

And yeah, that's how I imagined the stairs looked.

**Comment by gedads** [comment]
http://imgur.com/UeDXMnI there's something that need testing but after applying general settings, you can undo a tile and change settings and redo that tile only. it might be enough for stairs to be more accurate also it can work for small parts of the map where a mob can be stuck by a small edge etc.
I don't know if those mixed settings can work and also what's the maximum size for a navmesh. I'll try to ask lautan next time he's up.

**Comment by teschnei** [comment]
There are two main issues that I can think of with line of sight.
One, the raycast is from position_t to position_t, but the position_t represents where your feet are.  Thus if there is a small incline between you and your target, the ray may pass through the floor to get from your feet to your targets feet.  Simple solution, doing a raycast from an offset y position.
Two, the y position for mobs does not change as they move along inclines - the client just auto corrects the placement of their model.  So if a mob moves along an incline, their y position as reported by the server can be pretty different from what the client corrects it to - meaning the raycast can still be going through the floor.  This one's a bit harder to fix, it means clamping their y position to the navmesh plane they are moving along.  It may actually end up being easy to do, but I've never looked into the navmesh code before.

Either way, I can probably look into them this week when I have time.  The first one should be a quick fix

**Comment by teschnei** [comment]
this is all not including the possibility that the navmeshes are not perfectly in sync with the game world (as Hozu was seeing earlier)

**Comment by Hozu** [comment]
Clamping mob Y pos to the navmesh plane would fix being drawn into the ground as well, as you mentioned before.

**Comment by teschnei** [comment]
Yep, I forgot to mention that (though I was thinking it)

**Comment by teschnei** [comment]
Looks like the y pos clamping is already implemented.  Offset LOS I just implemented, so that's the y-value problems solved.
